base/reinterpretarray.jl

```diff
@@ -105,31 +175,132 @@ function check_writable(a::ReinterpretArray{T, N, S} where N) where {T,S}
     end
 end
 
-IndexStyle(a::ReinterpretArray) = IndexStyle(a.parent)
+## IndexStyle specializations
+
+# For `reinterpret(reshape, T, a)` where we're adding a channel dimension and with
+# `IndexStyle(a) == IndexLinear()`, it's advantageous to retain pseudo-linear indexing.
+struct IndexSCartesian2{K,N} <: IndexStyle end   # K = sizeof(S) ÷ sizeof(T), a static-sized 2d cartesian iterator
+
+IndexStyle(::Type{ReinterpretArray{T,N,S,A,false}}) where {T,N,S,A<:AbstractArray{S,N}} = IndexStyle(A)
+function IndexStyle(::Type{ReinterpretArray{T,N,S,A,true}}) where {T,N,S,A<:AbstractArray{S}}
+    if sizeof(T) < sizeof(S)
+        IndexStyle(A) === IndexLinear() && return IndexSCartesian2{sizeof(S) ÷ sizeof(T),N}()
+        return IndexCartesian()
+    end
+    return IndexStyle(A)
+end
+IndexStyle(::IndexSCartesian2{K,N}, ::IndexSCartesian2{K,N}) where {K,N} = IndexSCartesian2{K,N}()
+
+struct SCartesianIndex2{K,N} <: AbstractCartesianIndex{N}
```

Comment:
I'm unsure about the `N` parameter here. It felt slightly weird to have it be a subtype of generic `AbstractCartesianIndex`, and this does make it clear how many dimensions the particular `SCartesianIndex2` is standing in for. The downside is that this will force a lot of completely useless specialization.